File: 18. 4Sum/18. 4Sum.dart
class Solution {
  List<List<int>> fourSum(List<int> nums, int target) {
    nums.sort();
    List<List<int>> result = [];
    
    for (int a = 0; a < nums.length - 3; a++) {
      if (a > 0 && nums[a] == nums[a-1]) continue;
      for (int b = a + 1; b < nums.length - 2; b++) {
        if (b > a + 1 && nums[b] == nums[b-1]) continue;
        int c = b + 1;
        int d = nums.length - 1;
        while (c < d) {
          int sum = nums[a] + nums[b] + nums[c] + nums[d];
          if (sum == target) {
            result.add([nums[a], nums[b], nums[c], nums[d]]);
            while (c < d && nums[c] == nums[c+1]) c++;
            while (c < d && nums[d] == nums[d-1]) d--;
            c++; d--;
          } else if (sum < target) {
            c++;
          } else {
            d--;
          }
        }
      }
    }
    
    return result;
  }
}

File: 18. 4Sum/18. 4Sum.js
function fourSum(nums, target) {
    nums.sort((a, b) => a - b); // Sort the array
    const result = [];
    
    const n = nums.length;
    
    for (let i = 0; i < n - 3; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue; // Skip duplicates for i
        
        for (let j = i + 1; j < n - 2; j++) {
            if (j > i + 1 && nums[j] === nums[j - 1]) continue; // Skip duplicates for j
            
            let left = j + 1;
            let right = n - 1;
            
            while (left < right) {
                const sum = nums[i] + nums[j] + nums[left] + nums[right];
                
                if (sum === target) {
                    result.push([nums[i], nums[j], nums[left], nums[right]]);
                    
                    // Skip duplicates for left
                    while (left < right && nums[left] === nums[left + 1]) left++;
                    // Skip duplicates for right
                    while (left < right && nums[right] === nums[right - 1]) right--;
                    
                    left++;
                    right--;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }
    }
    
    return result;
}

// Example usage
console.log(fourSum([1,0,-1,0,-2,2], 0)); // Output: [[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]]
console.log(fourSum([2,2,2,2,2], 8));    // Output: [[2,2,2,2]]